core/deployment/src/main/java/io/quarkus/deployment/dev/DevModeMain.java

```diff
@@ -86,35 +86,34 @@ public void start() throws Exception {
                     }
                 }
             }
-            final PathList.Builder appRoots = PathList.builder();
-            Path p = Path.of(context.getApplicationRoot().getMain().getClassesPath());
-            if (Files.exists(p)) {
-                appRoots.add(p);
+            final PathList.Builder appRootsBuilder = PathList.builder();
+            final Path classesPath = Path.of(context.getApplicationRoot().getMain().getClassesPath());
+            if (Files.exists(classesPath)) {
+                appRootsBuilder.add(classesPath);
             }
             if (context.getApplicationRoot().getMain().getResourcesOutputPath() != null
                     && !context.getApplicationRoot().getMain().getResourcesOutputPath()
                             .equals(context.getApplicationRoot().getMain().getClassesPath())) {
-                p = Paths.get(context.getApplicationRoot().getMain().getResourcesOutputPath());
-                if (Files.exists(p)) {
-                    appRoots.add(p);
+                final Path resourcesOutputPath = Paths.get(context.getApplicationRoot().getMain().getResourcesOutputPath());
+                if (Files.exists(resourcesOutputPath)) {
+                    appRootsBuilder.add(resourcesOutputPath);
                 }
             }
 
+            final PathList appRoots = appRootsBuilder.build();
             QuarkusBootstrap.Builder bootstrapBuilder = QuarkusBootstrap.builder()
-                    .setApplicationRoot(appRoots.build())
+                    .setApplicationRoot(appRoots)
                     .setIsolateDeployment(true)
                     .setLocalProjectDiscovery(context.isLocalProjectDiscovery())
                     .addAdditionalDeploymentArchive(path)
                     .setBaseName(context.getBaseName())
                     .setMode(context.getMode());
             if (context.getDevModeRunnerJarFile() != null) {
                 bootstrapBuilder.setTargetDirectory(context.getDevModeRunnerJarFile().getParentFile().toPath());
+            } else if (context.getApplicationRoot().getTargetDir() != null) {
```

Comment:
@geoand maybe it would be more sensible to prioritize `context.getApplicationRoot().getTargetDir()` over `context.getDevModeRunnerJarFile().getParentFile().toPath()`. Thing is, that would change the existing behavior, so maybe better not risk it?

Reply:
It's probably best left as is